File: Ashutosh-Singh/DynamicProgramming/FibonacciNumber.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int fib(int n) {
        if (n == 0) return 0;
        if (n == 1) return 1;
        
        int prev1 = 0, prev2 = 1;
        for (int i = 2; i <= n; ++i) {
            int curr = prev1 + prev2;
            prev1 = prev2;
            prev2 = curr;
        }
        return prev2;
    }
};

int main() {
    Solution solution;
    int n;
    cout << "Enter the value of n: ";
    cin >> n;
    
    cout << "The " << n << "-th Fibonacci number is: " << solution.fib(n) << endl;
    return 0;
}

File: AditiKaushik/DP/fibonacciNumber.java
//509. Fibonacci Number

import java.util.Scanner;

class fibonacciNumber {
    public static int fib(int n) {
        if (n <= 1)
            return n;
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter position of Fibonacci Number: ");
        int n = sc.nextInt();
        System.out.println("Fibonacci number at position " + n + " is: " + fib(n));
        sc.close();
    }
}

//Time Complexity: O(n)
//Space Complexity: O(n)